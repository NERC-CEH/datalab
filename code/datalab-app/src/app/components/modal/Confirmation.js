import React from 'react';
import PropTypes from 'prop-types';
import DatalabModal from './DatalabModal';
import IconButton from '../common/control/IconButton';

const Confirmation = ({ title, body, onSubmit, onCancel }) => (
  <DatalabModal title={title} body={body}>
    <IconButton onClick={onCancel} icon="clear">
      No
    </IconButton>
    <IconButton danger onClick={onSubmit} icon="delete">
      Yes
    </IconButton>
  </DatalabModal>
);

Confirmation.propTypes = {
  title: PropTypes.string.isRequired,
  body: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

export default Confirmation;
