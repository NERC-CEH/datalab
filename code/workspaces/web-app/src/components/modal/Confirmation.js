import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContentText from '@material-ui/core/DialogContentText';
import IconButton from '../common/control/IconButton';

const bodyToContent = (body) => {
  // Handle writing the body as either an individual, or a list of elements.
  if (Array.isArray(body)) {
    const contentTexts = body.map((t, i) => (<DialogContentText key={i}>{t}</DialogContentText>));

    return (
      <DialogContent>
        {contentTexts}
      </DialogContent>
    );
  }

  return (
    <DialogContent>
      <DialogContentText>{body}</DialogContentText>
    </DialogContent>
  );
};

const Confirmation = ({ title, body, onSubmit, onCancel, confirmText = 'Confirm Deletion', confirmIcon = 'delete', danger = true }) => (
  <Dialog open={true} maxWidth="md">
    <DialogTitle>{title}</DialogTitle>
    {bodyToContent(body)}
    <DialogActions>
      <IconButton onClick={onCancel} icon="clear">Cancel</IconButton>
      <IconButton danger={danger} onClick={onSubmit} icon={confirmIcon} disabled={onSubmit === undefined}>{confirmText}</IconButton>
    </DialogActions>
  </Dialog>
);

Confirmation.propTypes = {
  title: PropTypes.string.isRequired,
  body: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.arrayOf(PropTypes.string).isRequired]),
  onSubmit: PropTypes.func,
  onCancel: PropTypes.func.isRequired,
};

export default Confirmation;
