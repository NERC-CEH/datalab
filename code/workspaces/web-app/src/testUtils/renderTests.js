import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider as MuiThemeProvider } from '@mui/material/styles';
import theme from '../theme';

const AllTheProviders = ({ children }) => (
    <MuiThemeProvider theme={theme}>
        {children}
    </MuiThemeProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
