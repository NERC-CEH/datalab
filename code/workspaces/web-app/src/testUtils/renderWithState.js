import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import promise from 'redux-promise-middleware';
import {
  PROMISE_TYPE_PENDING,
  PROMISE_TYPE_SUCCESS,
  PROMISE_TYPE_FAILURE,
} from '../actions/actionTypes';

const promiseTypeSuffixes = [PROMISE_TYPE_PENDING, PROMISE_TYPE_SUCCESS, PROMISE_TYPE_FAILURE];

const TEST_USER = 'test-user';

export const buildDefaultTestState = () => ({
  authentication: {
    identity: {
      sub: TEST_USER,
    },
    permissions: {
      fetching: { inProgress: false },
      value: [],
    },
  },
  projectUsers: {
    fetching: { inProgress: false },
    value: [
      { userId: TEST_USER, role: 'admin' },
    ],
  },
});

export const renderWithState = (
  initialState,
  ComponentToRender,
  props = {},
  renderOptions = {},
) => {
  const mockStore = configureStore([promise({ promiseTypeSuffixes })]);
  const store = mockStore(initialState);

  const wrapper = render(
    <Provider store={store}>
      <ComponentToRender {...props} />
    </Provider>,
    renderOptions,
  );

  return {
    wrapper,
    store,
  };
};
