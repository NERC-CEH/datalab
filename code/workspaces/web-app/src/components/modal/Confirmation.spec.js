import React from 'react';
import { render, screen, fireEvent } from '../../testUtils/renderTests';
import Confirmation from './Confirmation';

function shallowRender(props) {
  return render(<Confirmation {...props} />).container;
}

describe('Confirmation', () => {
  const onSubmitMock = jest.fn();
  const onCancelMock = jest.fn();

  const generateProps = () => ({
    title: 'Title',
    body: 'Body',
    onSubmit: onSubmitMock,
    onCancel: onCancelMock,
  });

  beforeEach(() => jest.resetAllMocks());

  it('creates correct snapshot for confirmation dialog', () => {
    const props = generateProps();

    shallowRender(props);

    expect(screen.getByRole('dialog')).toMatchSnapshot();
  });

  it('creates correct snapshot for confirmation dialog with array in body', () => {
    const props = {
      ...generateProps(),
      body: ['Body', 'Next line', 'Final line'],
    };

    shallowRender(props);

    expect(screen.getByRole('dialog')).toMatchSnapshot();
  });

  it('wires up cancel function correctly', () => {
    const props = generateProps();

    shallowRender(props);
    fireEvent.click(screen.getByText('Cancel'));

    expect(onCancelMock).toHaveBeenCalled();
  });

  it('wires up submit function correctly', () => {
    const props = generateProps();

    shallowRender(props);
    fireEvent.click(screen.getByText('Confirm Deletion'));

    expect(onSubmitMock).toHaveBeenCalled();
  });
});
