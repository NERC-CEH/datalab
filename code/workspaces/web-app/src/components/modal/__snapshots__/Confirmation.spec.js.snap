// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirmation creates correct snapshot for confirmation dialog 1`] = `
<div
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6"
    >
      Title
    </h2>
  </div>
  <div
    class="MuiDialogContent-root"
  >
    <p
      class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Body
    </p>
  </div>
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text IconButton-button-82"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root IconButton-icon-84"
        >
          clear
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text IconButton-buttonDanger-83"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Confirm Deletion
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root IconButton-icon-84"
        >
          delete
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`Confirmation creates correct snapshot for confirmation dialog with array in body 1`] = `
<div
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6"
    >
      Title
    </h2>
  </div>
  <div
    class="MuiDialogContent-root"
  >
    <p
      class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Body
    </p>
    <p
      class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Next line
    </p>
    <p
      class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Final line
    </p>
  </div>
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text IconButton-button-217"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root IconButton-icon-219"
        >
          clear
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text IconButton-buttonDanger-218"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Confirm Deletion
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root IconButton-icon-219"
        >
          delete
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
